sklearn/covariance/_shrunk_covariance.py

```diff
@@ -106,12 +106,16 @@ def shrunk_covariance(emp_cov, shrinkage=0.1):
 
     where `mu = trace(cov) / n_features`.
     """
-    emp_cov = check_array(emp_cov)
-    n_features = emp_cov.shape[0]
+    emp_cov = check_array(emp_cov, allow_nd=True)
+    n_features = emp_cov.shape[-1]
+    if not 0 <= shrinkage <= 1:
+        raise ValueError("shrinkage must be in [0, 1] (Got %d)" % shrinkage)
 
-    mu = np.trace(emp_cov) / n_features
     shrunk_cov = (1.0 - shrinkage) * emp_cov
-    shrunk_cov.flat[:: n_features + 1] += shrinkage * mu
+    mu = np.trace(emp_cov, axis1=-2, axis2=-1) / n_features
+    while mu.ndim != emp_cov.ndim:
+        mu = mu[..., np.newaxis]
```

Comment:
Using a `while` loop here looks odd. NumPy should offer some way of doing this. I think that `expand_dims` should do the trick since we can know which dimensions we want to extend.

Reply:
Something like this:

```suggestion
    mu = np.extend_dims(
        mu, axis=tuple(range(mu.dim, emp_cov.ndim))
    )
```